apps/heft/src/plugins/JestPlugin/jest-build-transform.ts

```diff
@@ -4,30 +4,56 @@
 import * as path from 'path';
 import { Path, FileSystem } from '@rushstack/node-core-library';
 import { InitialOptionsWithRootDir } from '@jest/types/build/Config';
+import { JestTypeScriptDataFile, IJestTypeScriptDataFileJson } from './JestTypeScriptDataFile';
+
+// This caches jest-typescript-data.json file contents.
+// Map from jestOptions.rootDir --> IJestTypeScriptDataFileJson
+const dataFileJsonCache: Map<string, IJestTypeScriptDataFileJson> = new Map();
 
 /**
  * This Jest transformer maps TS files under a 'src' folder to their compiled equivalent under 'lib'
  */
 export function process(src: string, filename: string, jestOptions: InitialOptionsWithRootDir): string {
+  let jestTypeScriptDataFile: IJestTypeScriptDataFileJson | undefined = dataFileJsonCache.get(
+    jestOptions.rootDir
+  );
+  if (jestTypeScriptDataFile === undefined) {
+    // Read jest-typescript-data.json, which is created by Heft's TypeScript plugin.  It tells us
+    // which emitted output folder to use for Jest.
+    jestTypeScriptDataFile = JestTypeScriptDataFile.loadForProject(jestOptions.rootDir);
+    dataFileJsonCache.set(jestOptions.rootDir, jestTypeScriptDataFile);
+  }
+
+  // Is the input file under the "src" folder?
   const srcFolder: string = path.join(jestOptions.rootDir, 'src');
+
   if (Path.isUnder(filename, srcFolder)) {
-    const fileBasename: string = path.basename(filename, path.extname(filename));
-    const srcRelativeFolderPath: string = path.dirname(path.relative(srcFolder, filename));
+    // Example: /path/to/project/src/folder1/folder2/Example.ts
+    const parsedFilename: path.ParsedPath = path.parse(filename);
+
+    // Example: folder1/folder2
+    const srcRelativeFolderPath: string = path.relative(srcFolder, parsedFilename.dir);
+
+    // Example: /path/to/project/lib/folder1/folder2/Example.js
     const libFilename: string = path.join(
       jestOptions.rootDir,
-      'lib',
+      jestTypeScriptDataFile.emitFolderPathForJest,
       srcRelativeFolderPath,
-      `${fileBasename}.js`
+      `${parsedFilename.name}.js`
     );
 
     try {
       return FileSystem.readFile(libFilename);
     } catch (error) {
-      if (!FileSystem.isNotExistError(error)) {
+      if (FileSystem.isNotExistError(error)) {
+        throw new Error(
+          'jest-build-transform: The expected transpiler output file does not exist:\n' + libFilename
+        );
+      } else {
         throw error;
       }
     }
+  } else {
+    throw new Error('jest-build-transform: The input path is not under the "src" folder:\n' + filename);
```

Comment:
@iclanton I made this a hard error because I cannot think of any valid cases where we would encounter a `.ts` file but NOT want to transpile it.  

Whereas silently returning the untranspiled source code results in quite confusing errors.